src/relations/hacluster.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+"""hacluster relation hooks and helpers."""
+
+import json
+import logging
+from hashlib import shake_128
+from ipaddress import IPv4Address, ip_address
+from typing import Optional, Union
+
+import ops
+
+HACLUSTER_RELATION_NAME = "ha"
+
+logger = logging.getLogger(__name__)
+
+
+class HACluster(ops.Object):
+    """Defines hacluster functunality."""
+
+    def __init__(self, charm: ops.CharmBase):
+        super().__init__(charm, HACLUSTER_RELATION_NAME)
+
+        self.charm = charm
+
+        self.framework.observe(
+            charm.on[HACLUSTER_RELATION_NAME].relation_changed, self._on_changed
+        )
+        self.framework.observe(charm.on.config_changed, self._on_changed)
+
+    @property
+    def relation(self) -> ops.Relation:
```

Comment:
```suggestion
    def relation(self) -> Optional[ops.Relation]:
```

Reply:
addressed in 9039b2a9c7aba39ea020615d09ce290173240896